### ISSUE TITLE
Deviant Art download stopped working

### ISSUE THREAD

**Issue by Chuckydawhg** [4pr0n/ripme#425]
It was working perfectly yesterday, now it isn't...haven't done anything different

![ripme-error](https://cloud.githubusercontent.com/assets/25056183/21848095/626d08e6-d853-11e6-8db9-64f9cec3a9d5.PNG)

It just fails every time I try to download

**Comment by rautamiekka** [comment]
Confirmed, fails on 1.4.1. I'll try an older version.

**Comment by rautamiekka** [comment]
Same failure on 1.4.0.

**Comment by rautamiekka** [comment]
Since 1.3.0 is actually 1.4.0 there's nothing to test, and even older ones should be pointless.

**Comment by Chuckydawhg** [comment]
Thank you very much for the speedy reply haha, hmm...so what do you think
the problem might be there?...has deviant art changed something on their
website?

It seems to only be a problem with deviant art :/

**Comment by rautamiekka** [comment]
Most likely the current change to the gallery and folder views; noticed how the pieces show up, many of them over-zoomed past the frame, and the pages are automatically appended as you browse ? That most likely is the cause. Personally I hate how the pics show up (not only that, but although you can CTRL+F and write something to search for something specific it doesn't help you at all, nothing shows up and when I first tried (haven't tried again) the page wouldn't even move), but the auto-loading is nice.

**Comment by Chuckydawhg** [comment]
Bummer....best of luck fixing the program! :D

**Comment by Norithiel** [comment]
Same issue, 1.4.1

**Comment by rogueralsha** [comment]
Noticed the same issue with my own scraping program. It's making AJAX calls as you scroll to this address:

    http://www.deviantart.com/dapi/v1/gallery/0?iid=572maa5c9fc4eeee0d5450833009f301c971-iy07v1ix-1.1&mp=1

The mp appears to be the offset, starting with 1, but it only calls it after the first scroll down, so I don't know if the first batch of data can be gotten this way. The iid is a page_id provided in-page by a javascript declaration that starts with:

`window.__initial_body_data=`

Except it's only the part up to the dash in the iid in the UIRL, Here's an example of the JSON data (specific field is requestid):

```
{
  "dapx": {
    "client": "dw",
    "daid": "cb55273de8b4cac138433f6f8450e1a7",
    "requestid": "572mcf21a8eab036b705428ac2f05769dd7b",
    "log_data": {
      "artist": 10514192
    },
    "eventid": "daweb:gallery::::pageview",
    "delay_init_event": false
  },
  "csrf": "WzAKEQIafQGpEqbT.ojvngu.XESYBX-HFyAUyMA7y6nBvYAvatYgDoh8ftHbyiobJBg",
  "hub": {
    "user": "obfuscated because it's probably specific to my account",
    "ts": "1484577678",
    "auth": "also obfuscated",
    "subs": [
      
    ],
    "endpoint": "https:\/\/hub.deviantart.net"
  }
}
```

I have yet to find a source for the part after the dash.

I'm currently testing a browser-extension-based approach, but it appears that as you scroll through the page the higher-up elements actually disappear, making catching them a bit of a challenge. Hope some of this helps, I'll post what solution I end up with if I get one working.

**Comment by Hrxn** [comment]
Interesting stuff @rogueralsha, thanks.

I tried something similar recently, although completely unrelated to Deviant Art (or comparable sites).

But it's quite a challenge to deal with scraping sites that dynamically load content via script.
I only know a bit of Beautiful Soup, but this unfortunately doesn't help here.

Do you use something like Fiddler or Charles for developing and testing scraping stuff?

**Comment by rogueralsha** [comment]
I use Chrome's built-in tools. I usually start by opening the developer console to the network tab, filtering to XHR, and then refreshing the page. I then scroll through the page, and it shows all the AJAX requests being made. From there I analyze the request URL and headers for anything that looks like a token or ID, then start searching the source code of the page for where it's coming from. I use a chrome extension to test any CSS selectors I use to grab individual elements, and raw RegExping when selectors or XPath can't get what I want (usually a combination for accuracy). 

Good new on this though, I just noticed that the new thumbnails each have this span around them:

`<span class="thumb wide" data-sigil="torpedo-thumb deviation-thumb-185430113 peekable" style="width: 491px; height: 275px;" href="http://gooloo0-o.deviantart.com/art/Claymore-Flora-185430113" collect_rid="1:185430113" data-deviationid="185430113" data-meta="3.3" data-super-img="http://img09.deviantart.net/311c/i/2010/311/4/c/claymore_flora_by_gooloo0_o-d32eesh.jpg" data-super-width="900" data-super-height="506" data-super-transparent="false" data-super-full-img="http://img09.deviantart.net/311c/i/2010/311/4/c/claymore_flora_by_gooloo0_o-d32eesh.jpg" data-super-full-width="900" data-super-full-height="506">`

Take note of the data-super-fill-img tag. I've only verified it against a few images so far, but so far that looks to be the direct URL of the full-size image. Normally you'd have to navigate to the image page and deal with the download button and its token, but this looks to just take you right to the full image. I'm going to test it against a bigger gallery to make sure. This doesn't help with catching all the images in the gallery, but it does make the download process one step easier as long as they leave that there.

**Comment by rautamiekka** [comment]
Starts sounding like that change wasn't merely for visitors, but to block RipMe for starters. With that description it wasn't developed in 1 day either.

**Comment by Hrxn** [comment]
BTW, has there been any change made lately to browsing DeviantArt without an account? Or any other noticeable site redesign? Because I haven't visited the site in a long time, so I wouldn't know any differences.
Asking because it could be relevant if any IDs or tokens used by the AJAX/XHR requests to the site are linked to an user account, and maybe are handled differently.

**Comment by rogueralsha** [comment]
Not that I know of. The two parts of the ID sent with this new AJAX call are composed of a long string that changes every page view, and a second string that seems to be infringementing somehow, as the first few digits stay the same for a period of time but the last few digits to change each time.

**Comment by metaprime** [comment]
Might not be to block rippers, though it has that effect. Could simply be to improve site performance as loading a bunch of elements will make page responsiveness slower and slower. Unloading elements to allow them to be GC'd would improve performance while scrolling.

**Comment by rautamiekka** [comment]
The change's unforgivable problem is drag-drop support for rearranging your gallery and favs is gone; unless you go manually edit each one, they'll have to stay where they are. Also it thinks you wanna fav your own stuff (uploads and favs alike) when drag-dropping.

**Comment by rautamiekka** [comment]
DA ripping could be needed very soon, as an artist is threatening with total delete.

**Comment by metaprime** [comment]
@rautamiekka Is it possible to use JDownloader to save the content before it's too late? (Assuming their DA downloader is in better condition than Ripme's)

**Comment by metaprime** [comment]
BTW #441 tries to fix things but apparently is having new problems.

**Comment by Hrxn** [comment]
Example URL, before it's to late? To test the different options available.

**Comment by rautamiekka** [comment]
I spoke too soon ([s]he cancelled the idea some time later, and when I checked his/her stuff I remembered there wasn't much to save), but it's not granted it'd be the only one. I figured I'd not disclose the artist for the better of both of us, so here's another unrelated user whose gallery doesn't go through RipMe (no gallery for that matter with the official releases, my own included) but JDownloader started working after updating (it was broken earlier as well):

http://roky320.deviantart.com/gallery/

**Comment by Hrxn** [comment]
Okay..


> 3,507 Deviations 

means 3507 files in the end, or can one item (deviation) be a small set of multiple images?
I haven't seen something like this yet, but I don't really use DeviantArt that much..

**Comment by rautamiekka** [comment]
^ Ya, over 3½k uploads, but when I pointed JD to the root of the gallery it said 16 targets can be fetched, and it downloaded the whole 16, nothing else. I counted, those 16 are the ones in the root of the gallery.

**Comment by Hrxn** [comment]
Okay, I tried a run with https://github.com/mikf/gallery-dl

Making a bit of progress here, downloaded 1200 items before exiting with an error 😉 

As https://github.com/mikf/gallery-dl/blob/master/gallery_dl/extractor/deviantart.py indicates, it seems to use the DeviantArt API to fetch all items in a gallery.

I tried it without any account data, so this might be the reason. Will investigate further.

**Comment by Hrxn** [comment]
FYI, it is fixed with the latest version of [mikf/gallery-dl](https://github.com/mikf/gallery-dl).

It's working fine.

**Comment by ph0enixph0enix** [comment]
Any chance of a fix for Ripme for DeviantArt? It is a great piece of work and I would absolutely want stick with Ripme...

**Comment by rautamiekka** [comment]
JDownloader works too, it gets updated and even has many more features, but unlike RipMe (and maybe others, haven't tested) it can't do recursive ripping, likely just an intentional limitation of the plugin the software uses; for example, my favs don't have anything in the favs root, but have many folders and JD says checking the link was an error.

**Comment by Wiiplay123** [comment]
Just updated to 1.4.6 from 1.2.13 because I didn't realise until just now that newer updates were out. I mostly use ripme for deviantArt stuff, any updates planned to fix this?

**Comment by ph0enixph0enix** [comment]
I would also be intereseted in a fix.

**Comment by metaprime** [comment]
@Wiiplay123 This is already set to the milestone for upcoming updates, but the ability to fix it is dependent on whether devs have time to dedicate to it. Unfortunately the devs are all volunteers and will tend to work on what interests them most, or seems easy and worthwhile.

**Comment by metaprime** [comment]
Fixed in #516 (version 1.4.15+)
This discussion is rather lengthy so there might be smaller issues I did not notice.
Please post follow-up issues as new issues.
